clang-tools-extra/clang-tidy/readability/RedundantCastingCheck.cpp

```diff
@@ -0,0 +1,254 @@
+//===--- RedundantCastingCheck.cpp - clang-tidy ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "RedundantCastingCheck.h"
+#include "../utils/FixItHintUtils.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/Lexer.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::readability {
+
+static bool areTypesEqual(QualType S, QualType D) {
+  if (S == D)
+    return true;
+
+  const auto *TS = S->getAs<TypedefType>();
+  const auto *TD = D->getAs<TypedefType>();
+  if (TS != TD)
+    return false;
+
+  QualType PtrS = S->getPointeeType();
+  QualType PtrD = D->getPointeeType();
+
+  if (!PtrS.isNull() && !PtrD.isNull())
+    return areTypesEqual(PtrS, PtrD);
+
+  const DeducedType *DT = S->getContainedDeducedType();
+  if (DT && DT->isDeduced())
+    return D == DT->getDeducedType();
+
+  return false;
+}
+
+static bool areTypesEqual(QualType TypeS, QualType TypeD,
+                          bool IgnoreTypeAliases) {
+  const QualType CTypeS = TypeS.getCanonicalType();
+  const QualType CTypeD = TypeD.getCanonicalType();
+  if (CTypeS != CTypeD)
+    return false;
+
+  return IgnoreTypeAliases || areTypesEqual(TypeS.getLocalUnqualifiedType(),
+                                            TypeD.getLocalUnqualifiedType());
+}
+
+static bool areBinaryOperatorOperandsTypesEqual(const Expr *E,
+                                                bool IgnoreTypeAliases) {
+  if (!E)
+    return false;
+  const Expr *WithoutImplicitAndParen = E->IgnoreParenImpCasts();
+  if (!WithoutImplicitAndParen)
+    return false;
+  if (const auto *B = dyn_cast<BinaryOperator>(WithoutImplicitAndParen)) {
+    const QualType Type = WithoutImplicitAndParen->getType();
+    if (Type.isNull())
+      return false;
+
+    const QualType NonReferenceType = Type.getNonReferenceType();
+    const QualType LHSType = B->getLHS()->IgnoreImplicit()->getType();
+    if (!LHSType.isNull() &&
+        !areTypesEqual(LHSType.getNonReferenceType(), NonReferenceType,
+                       IgnoreTypeAliases))
+      return true;
+    const QualType RHSType = B->getRHS()->IgnoreImplicit()->getType();
+    if (!RHSType.isNull() &&
+        !areTypesEqual(RHSType.getNonReferenceType(), NonReferenceType,
+                       IgnoreTypeAliases))
+      return true;
+  }
+  return false;
+}
```

Comment:
I think that the name of this function sounds misleading compared to what it actually does.
`areBinaryOperatorOperandsTypesEqual` sounds like you are comparing the types of the operands with each other, not with the return type of the binary operator.
An example would be: The function returns `true` when the return type of the binary operator and the type of LHS are not equal.

My idea is:
- rename function to `areBinaryOperatorOperandsTypesEqualToOperatorResultType` (which is almost a full sentence, but I couldn't think of anything better)
- invert return statements
- add `!` before calling `areBinaryOperatorOperandsTypesEqualToOperatorResultType` in ln 148
- slightly change the if conditions for the LHS and RHS types  (this is where I changed the behavior compared to what it is currently)
  - e.g., previously, the code would continue checking `RHSType`, even though `LHSType.isNull()` could be true

I checked that this works and ran the tests successfully with (and the negation of the condition where `areBinaryOperatorOperandsTypesEqualToOperatorResultType` is called):

```c++
static bool areBinaryOperatorOperandsTypesEqualToOperatorResultType(
    const Expr *E, bool IgnoreTypeAliases) {
  if (!E)
    return true;
  const Expr *WithoutImplicitAndParen = E->IgnoreParenImpCasts();
  if (!WithoutImplicitAndParen)
    return true;
  if (const auto *B = dyn_cast<BinaryOperator>(WithoutImplicitAndParen)) {
    const QualType Type = WithoutImplicitAndParen->getType();
    if (Type.isNull())
      return true;

    const QualType NonReferenceType = Type.getNonReferenceType();
    const QualType LHSType = B->getLHS()->IgnoreImplicit()->getType();
    if (LHSType.isNull() || !areTypesEqual(LHSType.getNonReferenceType(),
                                           NonReferenceType, IgnoreTypeAliases))
      return false;
    const QualType RHSType = B->getRHS()->IgnoreImplicit()->getType();
    if (RHSType.isNull() || !areTypesEqual(RHSType.getNonReferenceType(),
                                           NonReferenceType, IgnoreTypeAliases))
      return false;
  }
  return true;
}
```

IMO this function name and body work better, what do you think?

The first two checks for `nullptr` of `E` and `WithoutImplicitAndParen` will never trigger, I think? But keeping them won't hurt for the added safety.